btrace-instr/src/main/java/org/openjdk/btrace/instr/ClassCache.java

```diff
@@ -58,26 +93,40 @@ public ClassInfo get(ClassLoader cl, String className) {
     return get(cl, new ClassName(className));
   }
 
-  synchronized ClassInfo get(ClassLoader cl, ClassName className) {
-    Map<ClassName, ClassInfo> infos = getInfos(cl);
+  ClassInfo get(ClassLoader cl, ClassName className) {
+    ConcurrentMap<ClassName, ClassInfo> infos = getInfos(cl);
 
-    ClassInfo ci = infos.get(className);
-    if (ci == null) {
-      ci = new ClassInfo(this, cl, className);
-      infos.put(className, ci);
-    }
-    return ci;
+    return infos.computeIfAbsent(className, k -> new ClassInfo(ClassCache.this, cl, k));
   }
 
-  private synchronized Map<ClassName, ClassInfo> getInfos(ClassLoader cl) {
+  ConcurrentMap<ClassName, ClassInfo> getInfos(ClassLoader cl) {
     if (cl == null) {
       return bootstrapInfos;
     }
-    Map<ClassName, ClassInfo> infos = cacheMap.computeIfAbsent(cl, k -> new HashMap<>(500));
+    boolean[] rslt = new boolean[] {false};
+    ConcurrentMap<ClassName, ClassInfo> infos =
+        cacheMap.computeIfAbsent(
+            getClKey(cl),
+            k -> {
+              rslt[0] = true;
+              return new ConcurrentHashMap<>(500);
+            });
+    if (rslt[0]) {
+      ClassLoaderReference ref = new ClassLoaderReference(cl, cleanupQueue);
+      loaderRefs.put(ref, ref);
+    }
     return infos;
   }
 
+  private static String getClKey(ClassLoader cl) {
+    return cl.getClass().getName() + "@" + System.identityHashCode(cl);
```

Comment:
Think use Pair is better than concatenate string for key

Reply:
Yes, this wold be the right thing to do. Once I have your confirmation that this actually resolves your problems I will polish this PR.
Anyway, thanks for chiming in! Not a lot of feedback on PRs here usually ...